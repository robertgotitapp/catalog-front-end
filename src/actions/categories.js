import { CategoriesAction } from '../utils/const';
import { post, get } from '../utils/requests';

export function addCategory(accessToken, category) {
  return {
    type: CategoriesAction.ADD_CATEGORY,
    promise: post('http://127.0.0.1:5000/categories', accessToken, category),
  };
}

export function getCategories(offset, limit) {
  return {
    type: CategoriesAction.GET_CATEGORIES,
    promise: get(`http://127.0.0.1:5000/categories?offset=${offset}&limit=${limit}`),
  };
}

export function selectCurrentCategory(selectedCategory) {
  return {
    type: CategoriesAction.SELECT_CURRENT_CATEGORY,
    payload: { currentCategory: selectedCategory },
  };
}
