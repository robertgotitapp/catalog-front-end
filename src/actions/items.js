import { ItemsAction, HeadersType } from '../utils/const';
import {
  get, post, update, remove,
} from '../utils/requests';

export function addItem(categoryId, item) {
  return {
    type: ItemsAction.ADD_ITEM,
    promise: post(`http://127.0.0.1:5000/categories/${categoryId}/items`,
      [HeadersType.CONTENTTYPE, HeadersType.AUTHORIZATION],
      item),
  };
}

export function getItems(categoryId, offset, limit) {
  return {
    type: ItemsAction.GET_ITEMS,
    promise: get(`http://127.0.0.1:5000/categories/${categoryId}/items?offset=${offset}&limit=${limit}`),
  };
}

export function removeItem(categoryId, itemId) {
  return {
    type: ItemsAction.DELETE_ITEM,
    promise: remove(`http://127.0.0.1:5000/categories/${categoryId}/items/${itemId}`,
      [HeadersType.CONTENTTYPE, HeadersType.AUTHORIZATION]),
  };
}

export function updateItem(categoryId, itemId, item) {
  console.log(categoryId);
  console.log(item);
  return {
    type: ItemsAction.UPDATE_ITEM,
    promise: update(
      `http://127.0.0.1:5000/categories/${categoryId}/items/${itemId}`,
      [HeadersType.CONTENTTYPE, HeadersType.AUTHORIZATION],
      item,
    ),
  };
}
