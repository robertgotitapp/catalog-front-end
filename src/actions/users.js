export const ADD_USER = 'ADD_USER'
export const AUTHENTICATE_USER = 'AUTHENTICATE_USER'

export function addUser(user) {
    return {
        type: ADD_USER,
        user
    }
}

export function handleAddUser(user) {
    return (dispatch) => {
        dispatch(addUser(user))
    }
}

