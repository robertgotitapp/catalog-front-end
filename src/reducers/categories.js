import { CategoriesAction } from '../utils/const';

const initialState = { categories: {}, currentCategory: null };

export default function categoriesReducer(state = initialState, action) {
  switch (action.type) {
    case CategoriesAction.ADD_CATEGORY_SUCCESS:
      // fixing problem with id not consitent
      return {
        ...state,
        [action.payload.id]: action.payload,
      };
    case CategoriesAction.GET_CATEGORIES_SUCCESS:
      return {
        ...state,
        categories: action.payload.categories,
      };
    case CategoriesAction.SELECT_CURRENT_CATEGORY:
      return {
        ...state,
        currentCategory: action.payload.currentCategory,
      };
    default:
      return state;
  }
}
