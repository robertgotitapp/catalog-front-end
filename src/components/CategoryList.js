import React, { Component } from 'react';
import { connect } from 'react-redux';
import Card from 'react-bootstrap/Card';
import Nav from 'react-bootstrap/Nav';
import Tab from 'react-bootstrap/Tab';
import { getItems, selectItemPage } from '../actions/items';
import { selectCurrentCategory, getCategories } from '../actions/categories';
import { PaginationConfig } from '../utils/const';

export class CategoryList extends Component {
  state = { loading: true };

  componentDidMount() {
    this.props.getCategories(0, PaginationConfig.CATEGORIES_PER_PAGE)
      .then((res) => {
        if (res.statusCode) {
          this.setState({ loading: false });
        }
      });
  }

  viewCategory = (e) => {
    const categoryChosen = Number(e.target.name);
    this.props.selectCurrentCategory(categoryChosen);
    this.props.selectItemPage(PaginationConfig.DEFAULT_PAGE);
    this.props.getItems(categoryChosen,
      PaginationConfig.DEFAULT_OFFSET,
      PaginationConfig.ITEMS_PER_PAGE);
  };


  render() {
    if (this.state.loading === true) {
      return <div className='loader'> ...loading </div>;
    }

    return (
      <Card>
        <Card.Body>
          <Card.Title>Categories</Card.Title>
        </Card.Body>
        <Tab.Container>
          <Nav variant="pills" className="flex-column">
            { Object.keys(this.props.categories).map(key => (
              <Nav.Item key={key}>
                <Nav.Link eventKey={key} name={key} onClick={this.viewCategory}>
                  {this.props.categories[key].name}
                </Nav.Link>
              </Nav.Item>
            ))}
          </Nav>
        </Tab.Container>
      </Card>
    );
  }
}

function mapStateToProps({ categoriesReducer }) {
  return {
    categories: categoriesReducer.categories,
    currentCategory: categoriesReducer.currentCategory,
  };
}

const mapDispatchToProps = {
  getCategories,
  selectCurrentCategory,
  getItems,
  selectItemPage,
};

export default connect(mapStateToProps, mapDispatchToProps)(CategoryList);
