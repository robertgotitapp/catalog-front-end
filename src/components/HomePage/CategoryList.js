import React, { Component } from 'react';
import { connect } from 'react-redux';
import Card from 'react-bootstrap/Card';
import Nav from 'react-bootstrap/Nav';
import Tab from 'react-bootstrap/Tab';
import { getItems } from '../../actions/items';
import { selectCurrentCategory, getCategories } from '../../actions/categories';
import { PaginationConfig } from '../../utils/const';

export class CategoryList extends Component {
  componentDidMount() {
    this.props.selectCurrentCategory(this.props.categoryId);
    this.props.getItems(
      this.props.categoryId,
      this.props.pageNumber,
      PaginationConfig.ITEMS_PER_PAGE,
    );
  }

  viewCategory = (e) => {
    const categoryChosen = Number(e.target.name);
    this.props.selectCurrentCategory(categoryChosen);
    this.props.getItems(categoryChosen,
      PaginationConfig.DEFAULT_OFFSET,
      PaginationConfig.ITEMS_PER_PAGE);
    this.props.history.push(`/categories/${categoryChosen}/items/1`);
  };


  render() {
    return (
      <Card>
        <Card.Body>
          <Card.Title>Categories</Card.Title>
        </Card.Body>
        <Tab.Container>
          <Nav variant="pills" className="flex-column">
            { Object.keys(this.props.categories).map(key => (
              <Nav.Item key={key}>
                <Nav.Link
                  eventKey={key}
                  name={key}
                  onClick={this.viewCategory}
                >
                  {this.props.categories[key].name}
                </Nav.Link>
              </Nav.Item>
            ))}
          </Nav>
        </Tab.Container>
      </Card>
    );
  }
}

function mapStateToProps({ categories }, { categoryId, pageNumber }) {
  return {
    categories: categories.categories,
    currentCategory: categoryId,
    pageNumber,
    // currentCategory: categories.currentCategory,
  };
}

const mapDispatchToProps = {
  getCategories,
  selectCurrentCategory,
  getItems,
};

export default connect(mapStateToProps, mapDispatchToProps)(CategoryList);
