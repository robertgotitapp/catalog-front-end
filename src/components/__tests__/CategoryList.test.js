import React from 'react';
import { shallow } from 'enzyme';
import ListGroupItem from 'react-bootstrap/ListGroupItem';
import { CategoryList } from '../CategoryList';
import { getCategories } from '../../actions/categories';

describe('components/CategoryList', () => {
  const props = {
    categories: {
      1: {
        created: '2019-06-17T13:46:18+00:00',
        description: 'Laptop',
        id: 1,
        name: 'Laptop',
        updated: '2019-06-17T13:46:18+00:00',
      },
      2: {
        created: '2019-06-17T13:46:25+00:00',
        description: 'Books',
        id: 2,
        name: 'Books',
        updated: '2019-06-17T13:46:25+00:00',
      },
      3: {
        created: '2019-06-17T13:46:48+00:00',
        description: 'Speaker',
        id: 3,
        name: 'Speaker',
        updated: '2019-06-17T13:46:48+00:00',
      },
    },
    getCategories: jest.fn(() => Promise.resolve()),
  };
  const container = shallow(<CategoryList {...props} />);
  container.setState({ loading: false });

  it('should render correctly', () => {
    expect(container).toMatchSnapshot();
  });

  it('should contain exact number of categories', () => {
    expect(container.find(ListGroupItem)).toHaveLength(3);
  });
});
