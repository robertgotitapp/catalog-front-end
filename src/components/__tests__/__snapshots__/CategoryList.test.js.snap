// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/CategoryList should render correctly 1`] = `
<div>
  <Card
    body={false}
  >
    <CardBody>
      <CardTitle>
        Categories
      </CardTitle>
    </CardBody>
  </Card>
  <ForwardRef>
    <Bootstrap(ListGroupItem)
      key="1"
    >
      <Connect(Category)
        category={
          Object {
            "created": "2019-06-17T13:46:18+00:00",
            "description": "Laptop",
            "id": 1,
            "name": "Laptop",
            "updated": "2019-06-17T13:46:18+00:00",
          }
        }
      />
    </Bootstrap(ListGroupItem)>
    <Bootstrap(ListGroupItem)
      key="2"
    >
      <Connect(Category)
        category={
          Object {
            "created": "2019-06-17T13:46:25+00:00",
            "description": "Books",
            "id": 2,
            "name": "Books",
            "updated": "2019-06-17T13:46:25+00:00",
          }
        }
      />
    </Bootstrap(ListGroupItem)>
    <Bootstrap(ListGroupItem)
      key="3"
    >
      <Connect(Category)
        category={
          Object {
            "created": "2019-06-17T13:46:48+00:00",
            "description": "Speaker",
            "id": 3,
            "name": "Speaker",
            "updated": "2019-06-17T13:46:48+00:00",
          }
        }
      />
    </Bootstrap(ListGroupItem)>
  </ForwardRef>
</div>
`;
